### PR TITLE
admin panel

### PR DESCRIPTION
**Pull request by n4-t3**
Created admin panel to show the details about the class.

### CODE REVIEW

File: src/constants/ui.ts [hunk 1]
```diff
@@ -1,2 +1,3 @@
 export const MIN_BUTTON_SIZE = 48
+export const TABLET_WIDTH_THRESHOLD = 767
```
Comment (jhw123) [inline src/constants/ui.ts:2]: I think it is better to make another PR for Tablet UI supports

File: src/components/Gnb.tsx [hunk 2]
```diff
@@ -31,6 +36,7 @@ export const Gnb = () => {
           {/* <MenuBtn onClick={onClickMenu('/class/' + cid)}>Question List</MenuBtn> */}
           <MenuBtn onClick={onClickSwitchClass}>Switch Class</MenuBtn>
           <MenuBtn onClick={onClickMyPage}>My Page</MenuBtn>
+          {isAdmin && cid && <MenuBtn onClick={onClickAdmin}>Admin</MenuBtn>}
```
Comment (jhw123) [inline src/components/Gnb.tsx:39]:
```suggestion
          {isAdmin && <MenuBtn onClick={onClickAdmin}>Admin</MenuBtn>}
```
could you change `onClickAdmin` to check cid value instead?

File: src/components/Gnb.tsx [hunk 3]
```diff
@@ -31,6 +36,7 @@ export const Gnb = () => {
           {/* <MenuBtn onClick={onClickMenu('/class/' + cid)}>Question List</MenuBtn> */}
```
Comment (jhw123) [inline src/components/Gnb.tsx:36]: Let's remove this dead code

File: src/pages/admin/[cid]/index.tsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import { FillBtn } from '@components/basic/button/Button'
+
+export default function Page() {
+  const isAdmin = useSelector((state: RootState) => state.userInfo.isAdmin)
+  const { push, query } = useRouter()
+  const cid = query.cid as string
+  const [classInfo, setClassInfo] = useState<LoadClassInfoResults>()
+
+  useEffect(() => {
+    !isAdmin
```
Comment (jhw123) [inline src/pages/admin/[cid]/index.tsx:20]: I suggest to use `if` and `else` for better readability. Ternary operators are more appropriate when you need the return value.

File: src/pages/admin/[cid]/index.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+              <Col>{classInfo?.students.length}</Col>
+              <Col>
+                <FillBtn onClick={onClick}>Detail</FillBtn>
+              </Col>
+            </TableRow>
+            <TableRow>
+              <Col>Questions</Col>
+              <Col>{classInfo?.qstems.length}</Col>
+              <Col>
+                {' '}
```
Comment (jhw123) [inline src/pages/admin/[cid]/index.tsx:66]: is this space intended?

File: src/pages/admin/[cid]/index.tsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      : cid
+      ? request<LoadClassInfoParams, LoadClassInfoResults>(`admin/loadClassInfo`, { cid }).then(res => {
+          if (res) {
+            setClassInfo(res)
+          }
+        })
+      : null
+  }, [isAdmin, push, cid, setClassInfo])
+
+  const onClick = useCallback(() => {}, [])
```
Comment (jhw123) [inline src/pages/admin/[cid]/index.tsx:31]: why is this empty?

File: src/pages/admin/[cid]/index.tsx [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+const TableRow = styled.li`
+  background-color: ${palette.common.white};
+  box-shadow: 0px 0px 9px 0px ${palette.background.dark};
+`
+
+const Col = styled.div`
+  flex-basis: 33%;
+  @media all and (max-width: ${TABLET_WIDTH_THRESHOLD}px) {
+    flex-basis: 100%;
+    display: flex;
```
Comment (jhw123) [inline src/pages/admin/[cid]/index.tsx:126]: fyi, look into `display: grid` as well

File: src/pages/api/admin/loadClassInfo.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import { Types } from 'mongoose'
+
+export interface LoadClassInfoParams {
+  cid: string
+}
+
+export interface LoadClassInfoResults {
+  name: string
+  topics: string[]
+  students: Types.ObjectId[]
```
Comment (jhw123) [inline src/pages/api/admin/loadClassInfo.ts:12]: We are using `ID` type to refer to id values.

**Review by jhw123 — CHANGES_REQUESTED** [review]

File: src/components/Gnb.tsx (hunk 2 above)
Comment (n4-t3) [reply src/components/Gnb.tsx]: > could you change `onClickAdmin` to check cid value instead?

Alright:)

File: src/pages/admin/[cid]/index.tsx (hunk 5 above)
Comment (n4-t3) [reply src/pages/admin/[cid]/index.tsx]: > is this space intended?

It was accidental:)

File: src/pages/admin/[cid]/index.tsx (hunk 4 above)
Comment (n4-t3) [reply src/pages/admin/[cid]/index.tsx]: > I suggest to use `if` and `else` for better readability. Ternary operators are more appropriate when you need the return value.

Oh ok, I have updated it that way

File: src/components/Gnb.tsx (hunk 3 above)
Comment (n4-t3) [reply src/components/Gnb.tsx]: > Let's remove this dead code

removed:)

File: src/pages/admin/[cid]/index.tsx (hunk 6 above)
Comment (n4-t3) [reply src/pages/admin/[cid]/index.tsx]: > why is this empty?

I was planning to create another PR for different pages but I will just include them here:)

File: src/pages/api/admin/loadClassInfo.ts (hunk 8 above)
Comment (n4-t3) [reply src/pages/api/admin/loadClassInfo.ts:12]: > We are using `ID` type to refer to id values.

we are using this just to count the number of students in the class not for anything else so I hope it is fine:)

File: src/constants/ui.ts (hunk 1 above)
Comment (n4-t3) [reply src/constants/ui.ts:2]: > I think it is better to make another PR for Tablet UI supports

Ok I will create another PR for it:)

File: src/components/MadeOption.tsx [hunk 9]
```diff
@@ -23,6 +23,10 @@ export const MadeOption = (props: Props) => {
     setIsOpenModal(!isOpenModal)
   }, [setIsOpenModal, isOpenModal])
 
+  const closeModal = useCallback(() => {
```
Comment (jhw123) [inline src/components/MadeOption.tsx:25]: How are toggleModal and closeModal different? Is there any case that they are different?

File: src/pages/admin/[cid]/topics.tsx [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+import styled from '@emotion/styled'
+import { palette, typography } from '@styles/theme'
+import { TABLET_WIDTH_THRESHOLD } from 'src/constants/ui'
+import { FillBtn, TextBtnCta } from '@components/basic/button/Button'
+import { UpdateTopicInfoParams, UpdateTopicInfoResults } from '@api/admin/updateTopicInfo'
+import { UpdateTopicDialog } from '@components/Dialogs/updateTopicDialog'
+import { CheckDialog } from '@components/Dialogs/CheckDialog'
+
+export default function Page() {
+  const isAdmin = useSelector((state: RootState) => state.userInfo.isAdmin)
```
Comment (jhw123) [inline src/pages/admin/[cid]/topics.tsx:18]: Perhaps, a more secure way is to block users from entering the page. We can do this with server side props
https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#redirect

Comment (jhw123) [reply src/pages/admin/[cid]/topics.tsx:18]: we can change this later

File: src/pages/api/admin/loadClassInfo.ts (hunk 8 above)
Comment (jhw123) [reply src/pages/api/admin/loadClassInfo.ts:12]: I meant the type definition is wrong. If all you need is student count, it should be `number` instead of `List`

File: src/pages/api/admin/updateTopicInfo.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  topics: string[]
+}
+
+export interface UpdateTopicInfoResults {
+  topics: string[]
+  success: boolean
+}
+
+export default apiController<UpdateTopicInfoParams, UpdateTopicInfoResults>(async ({ cid, topics }) => {
+  const CourseClass = await ClassModel.findById(new Types.ObjectId(cid))
```
Comment (jhw123) [inline src/pages/api/admin/updateTopicInfo.ts:16]: PascalCase is used for Class names. We are using camelCase for variable names.

```suggestion
  const courseClass = await ClassModel.findById(new Types.ObjectId(cid))
```

File: src/pages/api/admin/updateTopicInfo.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  if (CourseClass) {
+    const updateCourseTopics = await CourseClass.updateOne({ $set: { topics: topics } })
+    if (updateCourseTopics) {
+      const CourseClass = await ClassModel.findById(new Types.ObjectId(cid))
+      return {
+        topics: CourseClass.topics,
+        success: true,
```
Comment (jhw123) [inline src/pages/api/admin/updateTopicInfo.ts:23]: cuz success is true no matter what, it seems meaningless. If you want to give success status, consider using `throw new Error()`

File: src/components/Gnb.tsx [hunk 13]
```diff
@@ -31,6 +40,7 @@ export const Gnb = () => {
...
           <MenuBtn onClick={onClickSwitchClass}>Classes</MenuBtn>
           <MenuBtn onClick={onClickMyPage}>My Page</MenuBtn>
+          {isAdmin && <MenuBtn onClick={onClickAdmin}>Admin</MenuBtn>}
```
Comment (jhw123) [inline src/components/Gnb.tsx:43]: I would not make the link to admin page visible (contained in client side code) on the client side for better security.

File: src/pages/api/admin/loadClassInfo.ts (hunk 8 above)
Comment (n4-t3) [reply src/pages/api/admin/loadClassInfo.ts]: > I meant the type definition is wrong. If all you need is student count, it should be `number` instead of `List`

Oh that's right updated it here https://github.com/kixlab/kuiz/pull/46/commits/b51043c3294b662d0fb4a7331b61915ff6839fc7

File: src/pages/api/admin/updateTopicInfo.ts (hunk 11 above)
Comment (n4-t3) [reply src/pages/api/admin/updateTopicInfo.ts]: > PascalCase is used for Class names. We are using camelCase for variable names.

updated it here https://github.com/kixlab/kuiz/pull/46/commits/2e04c862f97ba3b918c13396232055102a87d746

File: src/pages/api/admin/updateTopicInfo.ts (hunk 12 above)
Comment (n4-t3) [reply src/pages/api/admin/updateTopicInfo.ts]: > cuz success is true no matter what, it seems meaningless. If you want to give success status, consider using `throw new Error()`

That does make sense I have updated it here https://github.com/kixlab/kuiz/pull/46/commits/aa327e685ddb33ab773f0a3120a65a78808094da

File: src/pages/admin/[cid]/index.tsx (hunk 7 above)
Comment (n4-t3) [reply src/pages/admin/[cid]/index.tsx:134]: > fyi, look into `display: grid` as well

I didn't use `display:grid` but I have adapted responsive UI here https://github.com/kixlab/kuiz/pull/46/commits/cb5736217e5f7a3229022e4c72397187a576427d

File: src/components/MadeOption.tsx (hunk 9 above)
Comment (n4-t3) [reply src/components/MadeOption.tsx]: > How are toggleModal and closeModal different? Is there any case that they are different?

You are right; In this case it is redundant I have updated it here https://github.com/kixlab/kuiz/pull/46/commits/e1282d380f962fef67e3bab57c6f8bb3d4b5f734.

Comment (jhw123) [reply src/components/MadeOption.tsx]: I think CheckDialog should be also updated. For now, there will be (unfunctional) cancel button even if you are not passing closeModal prop.

File: src/pages/admin/[cid]/index.tsx (hunk 7 above)
Comment (jhw123) [reply src/pages/admin/[cid]/index.tsx:134]: I left some comments on the commit link

Comment (n4-t3) [reply src/pages/admin/[cid]/index.tsx:134]: > I left some comments on the commit link

I have updated them here https://github.com/kixlab/kuiz/pull/46/commits/e1282d380f962fef67e3bab57c6f8bb3d4b5f734

File: src/components/MadeOption.tsx (hunk 9 above)
Comment (n4-t3) [reply src/components/MadeOption.tsx]: > I think CheckDialog should be also updated. For now, there will be (unfunctional) cancel button even if you are not passing closeModal prop.

That is right; I have updated it here https://github.com/kixlab/kuiz/pull/46/commits/30e81da0c52c18618f82f6fc853513997d88e5b6

File: src/components/Gnb.tsx (hunk 13 above)
Comment (n4-t3) [reply src/components/Gnb.tsx:42]: > I would not make the link to admin page visible (contained in client side code) on the client side for better security.

I'm not really sure on how to tackle this problem 😅; what could be a better way to do it?

Comment (n4-t3) [reply src/components/Gnb.tsx]: > I would not make the link to admin page visible (contained in client side code) on the client side for better security.

is it something like this https://github.com/kixlab/kuiz/pull/46/commits/57a475225e818d6dc78160da1be139ae578d4f87?

Comment (jhw123) [reply src/components/Gnb.tsx]: oh I meant removing `<MenuBtn onClick={onClickAdmin}>Admin</MenuBtn>`. I think we can work on this in another PR.

**Review by jhw123 — APPROVED** [review]
